aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/compute/functions/InstanceToNodeMetadata.java

```diff
@@ -0,0 +1,136 @@
+   private final Supplier<Map<String, ? extends Image>> images;
+   private final Supplier<Map<String, ? extends Hardware>> hardwares;
+   private final Supplier<Set<? extends Location>> locations;
+   private final Function<Instance.Status, NodeMetadata.Status> toPortableStatus;
+   private final GroupNamingConvention groupNamingConvention;
+   @Resource @Named(ComputeServiceConstants.COMPUTE_LOGGER) protected Logger logger = Logger.NULL;
+
+   @Inject
+   public InstanceToNodeMetadata(InstanceTypeToHardware instanceTypeToHardware,
+                                 Supplier<Map<String, ? extends Image>> images,
+                                 Supplier<Map<String, ? extends Hardware>> hardwares,
+                                 @Memoized Supplier<Set<? extends Location>> locations,
+                                 Function<Instance.Status, NodeMetadata.Status> toPortableStatus,
+                                 GroupNamingConvention.Factory groupNamingConvention) {
+      this.instanceTypeToHardware = instanceTypeToHardware;
+      this.images = checkNotNull(images, "images cannot be null");
+      this.hardwares = checkNotNull(hardwares, "hardwares cannot be null");
+      this.locations = locations;
+      this.toPortableStatus = toPortableStatus;
+      this.groupNamingConvention = groupNamingConvention.createWithoutPrefix();
+   }
+
+   @Override
+   public NodeMetadata apply(Instance from) {
+      NodeMetadataBuilder builder = new NodeMetadataBuilder();
+
+      Optional<? extends Image> image = findImage(from.imageId());
+      if (image.isPresent()) {
+         builder.imageId(image.get().getId());
+         builder.operatingSystem(image.get().getOperatingSystem());
+      } else {
+         logger.info(">> image with id %s for instance %s was not found. "
+                     + "This might be because the image that was used to create the instance has a new id.",
+               from.instanceType(), from.instanceId());
+      }
+      Optional<? extends Hardware> hardware = findHardware(from.instanceType());
+      if (hardware.isPresent()) {
+         builder.hardware(hardware.get());
+      } else {
+         logger.info(">> hardware with id %s for instance %s was not found. "
+                         + "This might be because the image that was used to create the instance has a new id.",
+                 from.instanceType(), from.instanceId());
+      }
+
+      builder.id(RegionAndId.slashEncodeRegionAndId(from.regionId(), from.instanceId()));
+      builder.providerId(from.instanceId());
+      builder.name(from.instanceName());
+      builder.hostname(String.format("%s", from.hostname()));
+      builder.group(groupNamingConvention.extractGroup(from.instanceName()));
+      builder.status(toPortableStatus.apply(from.status()));
+      builder.privateAddresses(from.innerIpAddress().entrySet().iterator().next().getValue());
+      builder.publicAddresses(from.publicIpAddress().entrySet().iterator().next().getValue());
```

Comment:
Why do you just get the first IP and not all the IPs in the set?

Reply:
it's a weird data structure, getValue returns a List! :D